### PR TITLE
added --sites options for runTheMatrix

### LINKED ISSUE #22278: runTheMatrix.py -i all whitelists CERN even if files are elsewhere
Some workflows in runTheMatrix.py with the recycle option seek a GEN_SIM dataset that exists only at FNAL. The dasquery step fails because the command contains the requirement site=T2_CH_CERN. Could a command-line option --allsites, valid in conjunction with -i, be added to runTheMatrix.py that would allow the user of suppress this site requirement?

For example, in CMSSW_10_0_0_pre2 the command

runTheMatrix.py --command=--number=10 -w 2017 -l 10859.0 -i all

generates at step1
dasgoclient --limit 0 --query 'file dataset=/RelValQCD_Pt_3000_3500_13/CMSSW_10_0_0_pre2-100X_upgrade2018_realistic_v1-v1/GEN-SIM site=T2_CH_CERN'

But looking in DAS I find that this dataset exists only at T1_US_FNAL.

The query is generated in

Configuration/PyReleaseValidation/python/MatrixUtil.py

where we find

        if len(self.run) is not 0:
            return ["file {0}={1} run={2} site=T2_CH_CERN".format(query_by, query_source, query_run) for query_run in self.run]
            # return ["file {0}={1} run={2} ".format(query_by, query_source, query_run) for query_run in self.run]
        else:
            return ["file {0}={1} site=T2_CH_CERN".format(query_by, query_source)]
            # return ["file {0}={1} ".format(query_by, query_source)]

I can run successfully by reversing the comments in these two pairs of lines.

### PR DESCRIPTION
**Pull request by smuzaffar**
As requested here #22278 , added `--sites <site>` option to `runTheMatrix.py` script to select a specific site for recycle data. Setting it to emptry string will allow to search all sites.

This resolves #22278

### CODE REVIEW

**Comment by smuzaffar** [comment]
please test

**Comment by kpedro88** [comment]
+upgrade

File: Configuration/PyReleaseValidation/python/MatrixUtil.py [hunk 1]
```diff
@@ -177,11 +178,17 @@ def queries(self, dataset):
...
+        site = " site=T2_CH_CERN"
+        if "CMSSW_DAS_QUERY_SITES" in os.environ:
+            if os.environ["CMSSW_DAS_QUERY_SITES"]:
+                site = " site=%s" % os.environ["CMSSW_DAS_QUERY_SITES"]
+            else:
+                site = ""
         if len(self.run) is not 0:
-            return ["file {0}={1} run={2} site=T2_CH_CERN".format(query_by, query_source, query_run) for query_run in self.run]
+            return ["file {0}={1} run={2}{3}".format(query_by, query_source, query_run, site) for query_run in self.run]
             #return ["file {0}={1} run={2} ".format(query_by, query_source, query_run) for query_run in self.run]
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/MatrixUtil.py:189]: Do we still need to keep this line?

File: Configuration/PyReleaseValidation/python/MatrixUtil.py [hunk 2]
```diff
@@ -177,11 +178,17 @@ def queries(self, dataset):
...
         else:
-            return ["file {0}={1} site=T2_CH_CERN".format(query_by, query_source)]
+            return ["file {0}={1}{2}".format(query_by, query_source, site)]
             #return ["file {0}={1} ".format(query_by, query_source)]
```
Comment (chayanit) [inline Configuration/PyReleaseValidation/python/MatrixUtil.py:192]: Do we still need to keep this line?

**Comment by silviodonato** [comment]
merge
we can clean `MatrixUtil.py` in a follow up PR

**Comment by chayanit** [comment]
Hello @smuzaffar, @silviodonato  , this PR doesn't seem to remove a whitelist from CERN site. We have problem to read input for RelVals in 11_2_0_pre7. Can we revisit this?

**Comment by smuzaffar** [comment]
@chayanit , I did not want to change the default behavior of the script. So by default CERN site is selected. As mentioned int he description of the PR, running --site='' should allow to search all sites. If this behavior not working?

**Comment by chayanit** [comment]
@smuzaffar seems not, apparently it always looks for sample at CERN site and not others

**Comment by smuzaffar** [comment]
It works for me e.g running without --site option I see that runTheMatrix runs the following
```
>runTheMatrix.py -i all -l 1000 --maxSteps=1 --dryRun
dasgoclient --limit 0 --query 'file dataset=/MinimumBias/Run2011A-v1/RAW run=165121 site=T2_CH_CERN'
```

and with `--site=''` i get
```
>runTheMatrix.py -i all -l 1000 --maxSteps=1 --dryRun --site=''
 dasgoclient --limit 0 --query 'file dataset=/MinimumBias/Run2011A-v1/RAW run=165121'
```

**Comment by chayanit** [comment]
Hi @smuzaffar , it doesn't work for us and we found this is caused by the default setting here 
https://github.com/cms-sw/cmssw/blob/a65b92a2f0531f59ef024eb121979492591b52bd/Configuration/PyReleaseValidation/scripts/runTheMatrix.py#L287

**Comment by smuzaffar** [comment]
How can I reproduce it?

**Comment by chayanit** [comment]
Yes we run runTheMatrix to generate configuration

**Comment by smuzaffar** [comment]
Can you please share the runTheMatrix.py command you run?

**Comment by chayanit** [comment]
Sure, this is the command line we usually run for RelVal production
runTheMatrix.py --what upgrade -l 11834.0 -t 4 -m 11500 -b 'fullsim_PU_2021_14TeV' -i all --noCaf --wm force
while you need to update string here https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_steps.py#L3211 to "CMSSW_11_2_0_pre7-112X_mcRun3_2021_realistic_v8-v"

**Comment by smuzaffar** [comment]
I do not see `--sites=''` option here, so for sure it will use the default. As I have mentioned earlier that by default runTheMatrix.py uses `T2_CH_CERN` , this was the behavior for many year and I did not want to change that. That is why I added the extra command line option. In case you want to remove the `T2_CH_CERN` from the whitelist then please call runTheMatrix.py with `--sites=''` command-line option

**Comment by chayanit** [comment]
Ah ok @smuzaffar you mean we have to put --sites='' explicitly in the command line?

**Comment by smuzaffar** [comment]
yes

**Comment by chayanit** [comment]
Can you show how explicitly?

**Comment by smuzaffar** [comment]
```
runTheMatrix.py --what upgrade -l 11834.0 -t 4 -m 11500 -b 'fullsim_PU_2021_14TeV' -i all --noCaf --wm force --sites=''
```

**Comment by chayanit** [comment]
this option is not shown in runTheMatrix.py -h though should we include?

**Comment by smuzaffar** [comment]
are yo sure that you are using a release where this option is available? If I run `runTheMatrix.py --help` in CMSSW_11_2_X then I see
```
  --sites=DASSITES      Run DAS query to get data from a specific site
                        (default is T2_CH_CERN). Set it to empty string to
                        search all sites.
```

**Comment by chayanit** [comment]
Yeah now I saw the option is added to runTheMatrix.py script but it doesn't show up when run 'runTheMatrix.py --help' I'm doing it on 11_2_0_pre7

**Comment by davidlange6** [comment]
Seems to work for me

[dlange@lxplus757 dlange]$ scram p CMSSW_11_2_0_pre7
[dlange@lxplus757 dlange]$ cd CMSSW_11_2_0_pre7
[dlange@lxplus757 CMSSW_11_2_0_pre7]$ cmsenv
[dlange@lxplus757 CMSSW_11_2_0_pre7]$ runTheMatrix.py -h | grep -A3 site
  --ibeos               Use IB EOS site configuration
  --sites=DASSITES      Run DAS query to get data from a specific site
                        (default is T2_CH_CERN). Set it to empty string to
                        search all sites.

**Pull request merged**